fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/codec/TestFDBDirectory.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * TestFDBDirectory.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.record.lucene.LuceneEvents;
+import com.apple.foundationdb.record.lucene.LuceneIndexOptions;
+import com.apple.foundationdb.record.lucene.directory.FDBDirectory;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseFactory;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.ByteArrayUtil;
+import com.apple.foundationdb.tuple.Tuple;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.store.ByteBuffersDataInput;
+import org.apache.lucene.store.ByteBuffersIndexInput;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Wrapper for {@link FDBDirectory} to better support tests provided by lucene.
+ */
+public class TestFDBDirectory extends FDBDirectory {
+    private static final LuceneOptimizedFieldInfosFormat FIELD_INFOS_FORMAT = new LuceneOptimizedFieldInfosFormat();
+    private static boolean fullBufferToSurviveDeletes;
+    private static boolean allowAddIndexes;
+
+    private static AtomicReference<Pair<String, FieldInfos>> previousFieldInfos = new AtomicReference<>();
+    private static AtomicReference<Pair<String, Map<Long, byte[]>>> previousStoredFields = new AtomicReference<>();
+
+    public TestFDBDirectory() {
+        super(new Subspace(Tuple.from("record-test", "unit", "lucene")),
+                FDBDatabaseFactory.instance().getDatabase().openContext(),
+                Map.of(LuceneIndexOptions.OPTIMIZED_STORED_FIELDS_FORMAT_ENABLED, "true"));
```

Comment:
You may note that this does not include `PRIMARY_KEY_SEGMENT_INDEX_ENABLED`, that code broke with the tests because it tries to cast the `StoredFieldsReader`, but the reader is wrapped by the test framework.
https://github.com/FoundationDB/fdb-record-layer/blob/a1e93d2dc293877c8bb200783f5c5a81ef6ead1d/fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePrimaryKeySegmentIndex.java#L219

We may want to change the way that code works to work with the tests, but I think it makes sense to try and get the test framework in and then work to test it with more situations.

Reply:
https://github.com/FoundationDB/fdb-record-layer/issues/2407